src/relentless/optimize/objective.py

```diff
@@ -585,21 +585,22 @@ def _calc_ensemble_average_dvar_gradient(self, trajectory, variables):
                 filter_j_gt_i = neighbors[:, 1] > neighbors[:, 0]
                 neighbors.filter(filter_j_gt_i)
                 # pair contributions to the gradient
-                for i in self.potentials.pair.types:
-                    for j in self.potentials.pair.types:
-                        filter_ij = numpy.logical_and(
-                            type_masks[i][neighbors[:, 0]],
-                            type_masks[j][neighbors[:, 1]],
-                        )
-                        for var in variables:
-                            gradient[var] += numpy.sum(
-                                self.potentials.pair.derivative(
-                                    (i, j), var, x=neighbors.distances[filter_ij]
-                                )
+                if self.potentials.pair is not None:
+                    for i in self.potentials.pair.types:
+                        for j in self.potentials.pair.types:
+                            filter_ij = numpy.logical_and(
+                                type_masks[i][neighbors[:, 0]],
+                                type_masks[j][neighbors[:, 1]],
                             )
+                            for var in variables:
+                                gradient[var] += numpy.sum(
+                                    self.potentials.pair.derivative(
+                                        (i, j), var, x=neighbors.distances[filter_ij]
+                                    )
+                                )
 
                 # bond contributions to the gradient
-                if snap.bonds.N != 0:
+                if self.potentials.bond is not None:
```

Comment:
Do you want to do both checks? This code path only does something if both are true:
```suggestion
                if self.potentials.bond is not None and snap.bonds.N != 0:
```
Similarly for the angles and dihedrals